Credence game

ChickenProp opened the pull request:
Early implementation of the credence game. I don't expect it to be ready to go live, but I'd like feedback on what to do to make it ready.

moggyboy commented on credence-todo.org:
```diff
@@ -0,0 +1,30 @@
+
+* Possible feature: try to give users questions such that they end up with a reasonable distribution of reported credences.
+
+This is more complicated than just trying to give questions in categories they say they like. The two can presumably be combined.
```
I don't think this should live in the git repo, and if it does, please add it as a standalone markdown document in the root directory or append it to the README.md file.

moggyboy commented on db/questions/Databases.txt:
```diff
@@ -0,0 +1,5 @@
+Official CFAR Questions
+appliedrationality.org/uploads/games/credence/OfficialCfarQuestions.xml
+LessWrong Berlin Meetup Questions
+appliedrationality.org/uploads/games/credence/LW_Berlin.xml
```
Feels like this belongs in the README.md or a Creedence-game-specific markdown file as well.

moggyboy commented on db/schema.rb:
```diff
@@ -13,6 +13,53 @@
+    t.datetime "updated_at",                             :null => false
+    t.integer  "user_id"
+    t.integer  "num_answered",        :default => 0,     :null => false
+    t.boolean  "debug",               :default => false, :null => false
```
I don't love this. This is one of those things that people might scratch their heads over in a year's time wondering what it does. Can you either rename it to a more descriptive field name or remove it if it was only used during development?

moggyboy commented on db/schema.rb:
```diff
@@ -13,6 +13,53 @@
+    t.string   "prefix"
+    t.string   "suffix"
+    t.string   "type"
+    t.integer  "adjacentWithin"
```
Naming conventions dictate that this field should be named adjacent_within

moggyboy commented on config/routes.rb:
```diff
@@ -45,6 +45,10 @@
+  resource :credence,
+    :controller => 'credence',
+    :only => [:show, :update, :destroy]
+
```
This should be :credence_games, not :credence, and then you wouldn't have to specify a controller.

moggyboy commented on app/controllers/credence_controller.rb:
```diff
@@ -0,0 +1,56 @@
+    @title = "Credence game"
+
+    @game = CredenceGame.find_or_create_by_user_id current_user.id
+    @question = @game.current_question
```
This explodes if games haven't been populated yet, and probably should do something more sensible.

moggyboy commented:
In the course of testing this I fixed a bunch of things I considered to be a bit unclear in terms of naming and not consistent with naming conventions. There's a PR in the tricycle repo, but take a look at this commit:
https://github.com/tricycle/predictionbook/commit/072df977f6a028b9a0be02b6d7b1c95fa96c9174

... and let me know if there's anything you strongly object to. The only other thing I would like to see in *this* PR is a cleanup of the text files into a credence game specific markdown document and some Rspec feature specs that give coverage of the controller, views, etc.

ChickenProp commented:
No objection to your changes - I'll get some tests written at some point.
Flathunting is occupying much of my attention right now, so it might take
me a few weeks to get to it.

On Wed, Sep 16, 2015 at 6:15 AM Michael Webb <notifications@github.com>
wrote:

> In the course of testing this I fixed a bunch of things I considered to be
> a bit unclear in terms of naming and not consistent with naming
> conventions. There's a PR in the tricycle repo, but take a look at this
> commit:
> 072df97
> <https://github.com/tricycle/predictionbook/commit/072df977f6a028b9a0be02b6d7b1c95fa96c9174>
>
> ... and let me know if there's anything you strongly object to. The only
> other thing I would like to see in *this* PR is a cleanup of the text
> files into a credence game specific markdown document and some Rspec
> feature specs that give coverage of the controller, views, etc.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tricycle/predictionbook/pull/74#issuecomment-140628379>
> .
>

ChickenProp commented:
Okay, I'm looking into this now. Writing tests turns out to be a slow
process for me, as I'm not used to unit testing MVC apps or to the rails
ecosystem, but I'll get the hang of it. I wonder if you could quickly
comment on
https://github.com/ChickenProp/predictionbook/blob/credence-game/spec/views/credence_games/index.html.erb_spec.rb
?
In particular, I worry that having to stub the game id so that the view can
render, suggests I'm doing something wrong elsewhere. (I haven't actually
tested yet, but I don't think there's any risk if the user edits where the
form is submitted to - the controller takes the current user's existing
game, and doesn't look at any parameters. But it still seems unclean.)

On Thu, Sep 24, 2015 at 6:01 PM Philip Hazelden <philip.hazelden@gmail.com>
wrote:

> No objection to your changes - I'll get some tests written at some point.
> Flathunting is occupying much of my attention right now, so it might take
> me a few weeks to get to it.
>
>
> On Wed, Sep 16, 2015 at 6:15 AM Michael Webb <notifications@github.com>
> wrote:
>
>> In the course of testing this I fixed a bunch of things I considered to
>> be a bit unclear in terms of naming and not consistent with naming
>> conventions. There's a PR in the tricycle repo, but take a look at this
>> commit:
>> 072df97
>> <https://github.com/tricycle/predictionbook/commit/072df977f6a028b9a0be02b6d7b1c95fa96c9174>
>>
>> ... and let me know if there's anything you strongly object to. The only
>> other thing I would like to see in *this* PR is a cleanup of the text
>> files into a credence game specific markdown document and some Rspec
>> feature specs that give coverage of the controller, views, etc.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/tricycle/predictionbook/pull/74#issuecomment-140628379>
>> .
>>
>

moggyboy commented:
I think that the controller protects against a user overriding the game ID, so I don't think there's a problem. If I were you I would just focus on writing feature (integration) specs rather than unit testing (i.e. test the actual use cases for this functionality in the same way a real user would).

ChickenProp commented:
Okay! Sorry for the delay, I got completely distracted for a couple of
months, but I've picked it up again now. I've made a new pull request which
I think addresses everything you asked for. (Or at least, I've pushed to my
repo, and it seems to have updated the existing pull request.)

Something which I should probably do before this goes live, is add hooks
for updating existing questions. It's tricky because I don't know exactly
what kind of updates we'll want to do. The simple case is to change the
wording of a question to make it clearer or correct a typo, and that would
be easy enough. But if we want to add an answer (e.g. put Obama in 'which
president served longer?'), or we have a mistake in one, that seems
trickier. (If the rankings are wrong, then I think ideally we want to be
able to correct the rankings and adjust players' scores.)

I would suggest that most of this doesn't need to be solved yet, but that
hooks to make it solvable do need to happen early. My current thinking is
that the hooks probably just need to be short text ids in the XML files
that get imported to the db. Then any updates can reference that id without
having to look up the primary key in the db.

Does this seem reasonable to you? Do you know of a standard way to solve
this problem?

On Sun, Oct 18, 2015 at 10:03 PM Michael Webb <notifications@github.com>
wrote:

> I think that the controller protects against a user overriding the game
> ID, so I don't think there's a problem. If I were you I would just focus on
> writing feature (integration) specs rather than unit testing (i.e. test the
> actual use cases for this functionality in the same way a real user would).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tricycle/predictionbook/pull/74#issuecomment-149047265>
> .
>

moggyboy commented:
Few little quibbles, but this is starting to come together :+1:

ChickenProp commented:
Took a while again, but I think I addressed your comments.

On Thu, Jan 7, 2016 at 3:37 AM Michael Webb <notifications@github.com>
wrote:

> Few little quibbles, but this is starting to come together [image: :+1:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tricycle/predictionbook/pull/74#issuecomment-169539896>
> .
>

theminted commented on app/models/credence_game_response.rb:
```diff
@@ -0,0 +1,70 @@
+  belongs_to :second_answer, class_name: 'CredenceAnswer'
+
+  def self.pick_random
+    enabled_gens = CredenceQuestion.where(enabled: true)
```
Please avoid using abbreviations whenever possible. generations, genomes, genders instead?

theminted replied:
This still says `gens`. Please rename it to the full name of what it actually is, such as `generations`, `genders`, `genomes`. I do not know what a `gens` is.

theminted commented on app/models/credence_game_response.rb:
```diff
@@ -0,0 +1,70 @@
+      gen = enabled_gens.first(offset: rand(num_enabled))
+    end
+
+    gen.create_random_question
```
Same as above.

theminted replied:
This still says `gen`. I do not know what a `gen` is.

theminted commented on app/models/credence_game_response.rb:
```diff
@@ -0,0 +1,70 @@
+  end
+
+  def answer_correct?(ans)
+    ans == self.correct_index
```
abbreviation

theminted commented on app/models/credence_question.rb:
```diff
@@ -0,0 +1,62 @@
+      adjacent_within: gen['AdjacentWithin'].to_s.to_i,
+      weight: gen['Weight'].to_s.to_f
+    )
+    cq.save!
```
abbreviation

theminted commented on app/models/credence_question.rb:
```diff
@@ -0,0 +1,62 @@
+    rank = -1
+    last_val = nil
+    gen.search('Answer').each do |ans|
+      cur_val = ans['Value'].to_s
```
abbreviation

theminted commented on app/models/credence_statistics.rb:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    @intervals = Hash[ groups.map { |cred, group|
+      [ cred, Interval.new(cred, group.count, group.accuracy) ]
```
abbreviation

theminted commented on app/models/credence_statistics.rb:
```diff
@@ -0,0 +1,30 @@
+    groups = Hash.new { |h,k| h[k] = GroupedWagers.new }
+
+    questions.each { |q|
+      groups[q.answer_credence].add(q.answer_correct? q.given_answer)
```
abbreviation

theminted commented on app/models/credence_statistics.rb:
```diff
@@ -0,0 +1,30 @@
+class CredenceStatistics < BaseStatistics
+  def initialize(questions)
+    groups = Hash.new { |h,k| h[k] = GroupedWagers.new }
```
h, k, v for hash key value is fine. Well known abbreviations like i, j.

theminted commented on app/models/credence_question.rb:
```diff
@@ -0,0 +1,62 @@
+                                correct_index: which)
+  end
+
+  def self.create_from_element! (gen, id_prefix)
```
What does element refer to in this context? I don't see the name element used anywhere else in your PR. Maybe rename it to refer to one of the function arguments?

theminted commented on db/schema.rb:
```diff
@@ -11,7 +11,56 @@
+ActiveRecord::Schema.define(:version => 20160126231512) do
+
+  create_table "credence_answers", :force => true do |t|
+    t.integer  "credence_question_id"
```
This seems to be a foreign key referring to the table credence_questions. It's useful to index all foreign keys so you can do credence_question.credence_answers without having to scan through all records in the credence_answers table. This is usually not a big problem when the amount of SQL records is small, but as it grows it most likely will.

There are different schools on adding indexes. Some say "add indexes to foreign keys by default" while others say, "add no indexes and wait until performance issues start showing".

Please make your own choice for this scenario, and we'll go with that :)

ChickenProp replied:
I hadn't realized you could even have a non-indexed foreign key. Or is this just treated as an FK by rails, but not implemented with a MySQL FK? In any case, I had intended it for it to be indexed, and will add one accordingly.

theminted replied:
That's exactly it. Foreign keys are only SQL Integers in rails migrations. I'm not sure of the exact reason, but it might have something to do with compatibility across different types of SQL.

Awesome. Feel free to either add the index in the existing migrations that create the tables, or create new migrations that add them. Either works :)

theminted commented on spec/controllers/credence_games_controller_spec.rb:
```diff
@@ -0,0 +1,133 @@
+  describe 'Receiving a question' do
+    before(:each) do
+      CredenceQuestion.stub(:exists?).and_return(true)
+      @game = valid_credence_game
```
Please use let(:game) and subject(:game) to declare your variables.

ChickenProp replied:
I've updated to use let (in 0675fd1), but it's not clear to me where subject would be helpful. Could you give an example for this file?

I'm also not sure that let will improve speed in this case, because the before block is accessing the variables (and I don't think I can easily avoid that).

theminted commented on spec/models/credence_game_response_spec.rb:
```diff
@@ -0,0 +1,69 @@
+  end
+
+  it 'should give correct scores to specific credences' do
+    q = create_valid_credence_game_response(correct_index: 1)
```
abbreviations

theminted commented on spec/models/credence_game_response_spec.rb:
```diff
@@ -0,0 +1,69 @@
+
+describe CredenceGameResponse do
+  it 'should know right from wrong' do
+    q1 = create_valid_credence_game_response(correct_index: 0)
```
abbreviations

theminted commented on spec/models/credence_game_response_spec.rb:
```diff
@@ -0,0 +1,69 @@
+    expect(q1.answer_correct?(1)).to eq false
+    expect(q1.answer_correct?(0)).to eq true
+
+    q2 = create_valid_credence_game_response(correct_index: 1)
```
abbreviations

theminted commented on spec/models/credence_game_spec.rb:
```diff
@@ -0,0 +1,22 @@
+  include ModelFactory
+
+  it 'should calculate score correctly' do
+    g = valid_credence_game(num_answered: 3, score: 21)
```
abbreviations

theminted replied:
Is this supposed to be "game" perhaps?

theminted commented on spec/models/credence_question_spec.rb:
```diff
@@ -0,0 +1,92 @@
+    expect(cq.adjacent_within).to eq -1
+    expect(cq.weight).to eq 0.5
+
+    a0 = cq.credence_answers[0]
```
abbreviations

theminted commented on spec/models/credence_question_spec.rb:
```diff
@@ -0,0 +1,92 @@
+    expect(a0.text).to eq "first"
+    expect(a0.value).to eq "B"
+
+    a1 = cq.credence_answers[1]
```
abbreviations

theminted commented on spec/models/credence_game_response_spec.rb:
```diff
@@ -0,0 +1,69 @@
+    def make_generator (e=false)
+      gen = create_valid_credence_question(enabled: e)
+      a1 = create_valid_credence_answer(credence_question: gen, rank: 0)
+      a2 = create_valid_credence_answer(credence_question: gen, rank: 1)
```
`a1` and `a2` is assigned but never used

theminted commented on spec/models/credence_question_spec.rb:
```diff
@@ -0,0 +1,92 @@
+describe CredenceQuestion do
+  it 'should be able to create random questions' do
+    gen = create_valid_credence_question
+    as = (0..9).map do |rank|
```
abbreviations

theminted commented on spec/models/credence_question_spec.rb:
```diff
@@ -0,0 +1,92 @@
+    # gen.create_random_question is sufficiently slow that we don't want to do
+    # it loads of times. But if we don't do it enough, our test will be prone to
+    # failing randomly.
+    #   Is it possible to only have this test run if we request it explicitly?
```
No. If a spec is failing randomly, that's a sign that there are problems with your code. Rather than discarding the spec, it's better to fix the code so that the spec no longer fails randomly.

ChickenProp replied:
The problem here is the code is inherently random, and I want to run tests on the distribution it generates. By running it enough times we can make the probability of a false failure arbitrarily low, but then the test is slow.

It's possible that this sort of thing just shouldn't be a unit test, but I'm not sure what would be preferable.

theminted replied:
I see. That is an interesting scenario I have not encountered yet in testing.

I assume that the reason create_random_question is slow, is due to an active record being created every time the function is called. Here's what I would do. Separate the logic that generates the samples within the probability distribution into its own class like: SomeDistributionSampler with a .sample instance method. Then run the unit tests that checks the accuracy of the distribution on that class. If it's only simple arithmetic, you should be able to sample 10000+ before the tests start getting slow. Then use that distribution class in create_random_question and test that when you call create_random_question, SomeDistributionSampler gets a .sample call:

`expect_any_instance_of(SomeDistributionSampler).to receive(:sample).with('arguments', 'if', 'necessary')`

theminted commented on spec/models/credence_question_spec.rb:
```diff
@@ -0,0 +1,92 @@
+      counts[key] += 1
+    end
+
+    # What tests do we apply to counts?
```
Hmm.. not sure? :)

theminted commented:
A few nits with naming and structure, but generally seems pretty good. I'm not that familiar with the workings of the credence game, so my comments are mainly focused on general ruby and rails practices.

If you know someone else who is familiar with rails and the credence game, I'd definitely have them give this a quick look.

I'd say it's close to be ready for a deploy :+1: :sparkles:

ChickenProp commented:
I think I've addressed all your comments except the random unit test - I'll look at that next.

ChickenProp commented:
Got that one too - I did it with mocking instead of with a separate class.

theminted commented on app/models/base_statistics.rb:
```diff
@@ -0,0 +1,85 @@
+    attr_reader :heading, :count
+    def initialize(band, count=0, acc=0)
+      @heading = "#{band}%"
+      @acc = acc
```
accuracy

theminted commented on app/models/credence_answer.rb:
```diff
@@ -0,0 +1,9 @@
+
+  def format
+    # Would be nice to format the text in bold.
+    gen = self.credence_question
```
What is gen? Please use a full name.

Also, how is a gen the same thing as a credence question? That is what this line states.

ChickenProp replied:
The background here is that CredenceQuestion used to be named CredenceQuestionGenerator, so gen was short for generator. We renamed the class without renaming the variables, which I realize is confusing.

theminted commented on app/models/credence_game.rb:
```diff
@@ -0,0 +1,51 @@
+
+  def average_score
+    a = self.score.to_f / self.num_answered
+    a.finite? ? a : 0
```
What is a? Please use a full name

theminted commented on app/models/credence_game_response.rb:
```diff
@@ -0,0 +1,70 @@
+
+  def self.pick_random
+    enabled_questions = CredenceQuestion.where(enabled: true)
+    num_enabled = enabled_questions.count
```
num_enabled reads like: "The option num is enabled by being set to true"

I would avoid putting this in a temporary variable and instead just do the following below:
```
    question = enabled_questions.first(offset: rand(enabled_questions.count))
     while !question.enabled || question.weight < rand
       question = enabled_questions.first(offset: rand(enabled_questions.count))
     end
```

All you are accomplishing by putting this in a temp is to make the code harder to read.

theminted commented on app/models/credence_game_response.rb:
```diff
@@ -0,0 +1,70 @@
+  end
+
+  def score
+    _, s = self.score_answer(self.given_answer, self.answer_credence)
```
Please avoid single letter variable names.

ChickenProp replied:
How do you feel about the _ here? In my lexicon it's the standard name for a variable that never gets used, but I don't know if that's common in ruby.

theminted commented on spec/models/credence_game_spec.rb:
```diff
@@ -0,0 +1,22 @@
+
+  it 'should find answered questions' do
+    g = create_valid_credence_game
+    q1 = valid_answered_credence_question(credence_game: g)
```
question=

theminted commented:
There's still a few single letter variable names and abbreviations that need fixing, but other than that, this looks great! :sparkles: 👍

moggyboy commented:
👍  This looks good. I'm going to do a little prep work and get some dependencies updated, then I'll merge and deploy this.

moggyboy commented:
I had to do a lot of work to get this into a state I was happy with, but it's merged into master now. Will be deploying shortly along with a tonne of other updates.

moggyboy commented:
@ChickenProp I launched this but got reports of some answers being the wrong way around i.e. that the correct_index was set incorrectly for a response. I'm struggling to understand how the algorithm actually determines which answer is correct. Could you look over the code currently in master and point me in the right direction?

ChickenProp commented:
Ah crap, that's not good.

So each CredenceAnswer has a rank attribute, and whichever has the lowest rank is the correct answer. So one thing to check is that Warren Buffet should have a lower rank than Paul Allen. The rank is just a question of where it is in the XML file - the first Answer element becomes a CredenceAnswer with rank 0, the second gets rank 1, and so on; except that if two answers have the same value, they get the same rank and should never be compared to each other.

A CredenceGameResponse points to its two possible answers, and also has an attribute for which is the correct answer. That gets calculated in CredenceQuestion.create_random_response.

It also might be simply giving the wrong message, which is calculated in CredenceGameResponse.answer_message.

I'm not sure how any of these could be going wrong, but those are the places I'd look.

Separately, I think I did notice one bug in the code. In CredenceGameResponsesController.update, we check whether `@game.current_response_id == @response.id`, because if the user clicks multiple times then the server will take the second one as a response to a CredenceResponse they never had a chance to see, so we need to ignore it. But we also need to go back into the history, and look for the CredenceResponse they were looking at, to give the correct reply. It used to do that, but it looks like that got lost in the merge. I should have written a test for that, sorry. I don't think it could be what's going on here, though I'm not completely certain about that.
